skr/webhook/internal/handler.go

```diff
@@ -0,0 +1,343 @@
+		if !validatedResource.allowed {
+			return validatedResource
+		}
+
+		// new object
+		objectWatched := ObjectWatched{}
+		validatedResource = h.unmarshalRawObj(writer, admissionReview.Request.Object.Raw,
+			&objectWatched, resource.GroupVersionKind.String())
+		if !validatedResource.allowed {
+			return validatedResource
+		}
+
+		if resource.Status {
+			if !reflect.DeepEqual(oldObjectWatched.Status, objectWatched.Status) {
+				message = "sent requests to KCP for Status"
+			}
+		}
+
+		if resource.Spec && message == "" {
+			if !reflect.DeepEqual(oldObjectWatched.Spec, objectWatched.Spec) {
+				message = "sent requests to KCP for Spec"
+			}
+		}
+
+		if message != "" {
+			status = h.sendRequestToKcp(kymaName, objectWatched)
+		}
+
+		// since resource was found - exit loop
+		break
+	}
+
+	// send valid admission response - under all circumstances!
+	return h.validAdmissionReviewObj(message, status)
+}
+
+func (h *Handler) sendRequestToKcp(kymaName string, watched ObjectWatched) string {
+	// send request to kcp
+	watcherEvent := &types.WatcherEvent{
+		KymaCr:    kymaName,
+		Namespace: watched.Namespace,
+		Name:      watched.Name,
+	}
+	postBody, err := json.Marshal(watcherEvent)
+	if err != nil {
+		h.Logger.Error(err, "")
+		return metav1.StatusFailure
+	}
+
+	responseBody := bytes.NewBuffer(postBody)
+
+	kcpIP := os.Getenv("KCP_IP")
+	kcpPort := os.Getenv("KCP_PORT")
+	contract := os.Getenv("KCP_CONTRACT")
+	component := "manifest"
+
+	if kcpIP == "" || kcpPort == "" || contract == "" {
+		return metav1.StatusSuccess
+	}
+
+	url := fmt.Sprintf("http://%s/%s/%s/%s", net.JoinHostPort(kcpIP, kcpPort),
+		contract, component, config.EventEndpoint)
+
+	h.Logger.Info("KCP", "url", url)
+	//nolint:gosec
+	resp, err := http.Post(url, "application/json", responseBody)
+	if err != nil {
+		h.Logger.Error(err, "")
+		return metav1.StatusFailure
+	}
+
+	var response interface{}
+	if err = yaml.NewYAMLOrJSONDecoder(resp.Body, defaultBufferSize).Decode(response); err != nil {
+		h.Logger.Error(err, "response from KCP could not be unmarshalled")
+		return metav1.StatusFailure
+	}
+	h.Logger.Info("response", response)
+	return metav1.StatusSuccess
+}
+
+func (h *Handler) unmarshalRawObj(writer http.ResponseWriter, rawBytes []byte, response responseInterface,
+	resourceKind string,
+) validateResource {
+	if err := json.Unmarshal(rawBytes, response); err != nil || response.isEmpty() {
+		return h.invalidAdmissionReviewObj(writer, resourceKind, err)
+	}
+	return h.validAdmissionReviewObj("", "")
+}
+
+func (h *Handler) getKymaAndResourceListFromConfigMap(ctx context.Context) (*unstructured.Unstructured, []Resource) {
+	// fetch resource mapping ConfigMap
+	configMap := v1.ConfigMap{}
+	err := h.Client.Get(ctx, client.ObjectKey{
+		Name:      "skr-webhook-resource-mapping",
+		Namespace: "default",
+	}, &configMap)
+	if err != nil {
+		h.Logger.Error(err, "could not fetch resource mapping ConfigMap")
```

Comment:
if the configMap is nil here because err != nil then the next call to configMap.Data will cause a Nilpointer as configMap is nil

Reply:
I think this also needs retry / resilience logic

Reply:
it should throw a notfound error in that case

Reply:
https://github.com/kyma-project/kyma-watcher/issues/42